### PR TITLE
CSHARP-4615: Switch to polling monitoring on FaaS environment.

### PR DESCRIPTION
**Pull request by adelinowona**
- Adds serverMonitoringMode (stream|poll|auto) to URI options and MongoClientOptions
- Updates monitor behaviour to poll or stream based on options or environment.
- Ensures no dedicated connection is used for RTT pinging in polling mode.
- Updates Unified test runner to work with schema 1.17
- Syncs SDAM unified tests
- Syncs URI options tests

### Highlight

For users that want to control the behaviour of the server monitoring connection between each node in the topology, a new option, `serverMonitoringMode`, has been added. This defaults to auto but can be forced into a specific mode by providing a value of poll or stream. A polling monitor periodically issues a hello command to the node at an interval of `heartbeatFrequencyMS`. A streaming monitor sends an initial hello and then will automatically get a response from the Node when a change in server configuration occurs or at a maximum time of `heartbeatFrequencyMS`

### CODE REVIEW

File: src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs [hunk 1]
```diff
@@ -203,17 +210,25 @@ private CommandWireProtocol<BsonDocument> InitializeHelloProtocol(IConnection co
         {
             var connection = _connectionFactory.CreateConnection(_serverId, _endPoint);
 
+            _eventLoggerSdam.LogAndPublish(new ServerHeartbeatStartedEvent(connection.ConnectionId, _streamingEnabled && _currentDescription.TopologyVersion != null));
```
Comment (adelinowona) [inline src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs:213]: Adding this here as the spec tests seemed to require that the first server heartbeat events have the awaited field as false. Before this change we were only started logging server heartbeat events after the initial handshake (by then we know if the server supports streaming thus the awaited field would not be false causing spec tests to fail). I don't know if that was a requirement but I couldn't find anything about not emitting heartbeat events in the initial handshake in the specs. 

TLDR; I added the logging of server heartbeats when we initialize the server monitor connection as it is technically a heartbeat event.

File: tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs [hunk 2]
```diff
@@ -163,8 +161,8 @@ public void Description_should_be_connected_after_successful_heartbeat()
             subject.Description.State.Should().Be(ServerState.Connected);
             subject.Description.Type.Should().Be(ServerType.Standalone);
 
-            // no ServerHeartbeat events should be triggered during initial handshake
```
Comment (adelinowona) [inline tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs:166]: Is this driver's requirement or just one set for the C# driver? I couldn't find reasoning for this in the specs.

File: tests/MongoDB.Driver.Tests/UnifiedTestOperations/Matchers/UnifiedEventMatcher.cs [hunk 3]
```diff
@@ -371,6 +410,30 @@ private string GetAssertionErrorMessage(List<object> actualEvents, BsonArray exp
...
+                        break;
+                    case ServerHeartbeatSucceededEvent serverHeartbeatSucceededEvent:
+                        var serverHeartbeatSucceededEventDocument = new BsonDocument
+                        {
+                            { "awaited", serverHeartbeatSucceededEvent.Awaited }
+                        };
+                        actualEventsDocuments.Add(new BsonDocument(actualEvent.GetType().Name,
+                            serverHeartbeatSucceededEventDocument));
+                        break;
+                    case ServerHeartbeatFailedEvent serverHeartbeatFailedEvent:
```
Comment (adelinowona) [inline tests/MongoDB.Driver.Tests/UnifiedTestOperations/Matchers/UnifiedEventMatcher.cs:429]: These cases are very similar to each other and thought of collapsing them to the same together but I don't know how to do that while keeping the captured variable like:
case Event1:
case Event2:
case Event3 variable:
   variable is either of type Event1, Event2, Event3.

File: src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs [hunk 4]
```diff
@@ -48,6 +48,7 @@ internal sealed class ServerMonitor : IServerMonitor
         private readonly ServerMonitorSettings _serverMonitorSettings;
 
         private Thread _serverMonitorThread;
+        private bool _streamingEnabled;
```
Comment (BorisDog) [inline src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs:51]: minor: `_isStreamingEnabled`

File: src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs (hunk 1 above)
Comment (BorisDog) [reply src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs:213]: Yes this part is not consistent among drivers, and is being clarified in [DRIVERS-2677](https://jira.mongodb.org/browse/DRIVERS-2677).
This also should affect logging tests.
I would suggest to temporary adapt the tests accordingly, and make the code change if needed in [CSHARP-4831](https://jira.mongodb.org/browse/CSHARP-4831). Let's discuss more offline if needed.

File: src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs [hunk 5]
```diff
@@ -405,13 +420,19 @@ private void Heartbeat(CancellationToken cancellationToken)
...
                 }
 
-                processAnother =
-                    // serverSupportsStreaming
-                    (newDescription.Type != ServerType.Unknown && heartbeatHelloResult != null && heartbeatHelloResult.TopologyVersion != null) ||
-                    // connectionIsStreaming
-                    (helloProtocol != null && helloProtocol.MoreToCome) ||
-                    // transitionedWithNetworkError
+                var serverSupportsStreaming = (newDescription.Type != ServerType.Unknown &&
+                                               heartbeatHelloResult != null &&
```
Comment (BorisDog) [inline src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs:424]: formatting: new lines alignment (no need for parenthesis) 
```
var serverSupportsStreaming = newDescription.Type != ServerType.Unknown &&
   heartbeatHelloResult != null &&
   ...
```

File: src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs [hunk 6]
```diff
@@ -471,6 +498,16 @@ private void SetDescription(ServerDescription oldDescription, ServerDescription
             OnDescriptionChanged(oldDescription, newDescription);
         }
 
+        private bool IsRunningInFaaS()
+        {
+            return (Environment.GetEnvironmentVariable("AWS_EXECUTION_ENV")?.StartsWith("AWS_Lambda_") ?? false) ||
```
Comment (BorisDog) [inline src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs:503]: I could not find this in spec PR, is this specified in spec somewhere?

**Review by BorisDog — CHANGES_REQUESTED** [review]
Partial review, looks great so far.

File: src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs [hunk 7]
```diff
@@ -427,22 +448,28 @@ bool IsNetworkError(Exception ex)
...
 
             var stopwatch = Stopwatch.StartNew();
             try
             {
                 var helloResult = HelloHelper.GetResult(connection, helloProtocol, cancellationToken);
                 stopwatch.Stop();
 
-                _eventLoggerSdam.LogAndPublish(new ServerHeartbeatSucceededEvent(connection.ConnectionId, stopwatch.Elapsed, connection.Description.HelloResult.TopologyVersion != null, helloResult.Wrapped));
+                // RTT check if using polling monitoring
+                if ((!_streamingEnabled || (_streamingEnabled && connection.Description.HelloResult.TopologyVersion == null)) && helloResult != null)
```
Comment (BorisDog) [inline src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs:460]: Equivalent to shorter  `(!_streamingEnabled || connection.Description.HelloResult.TopologyVersion == null)` ?

File: tests/MongoDB.Driver.Core.Tests/Core/Configuration/ServerSettingsTests.cs [hunk 8]
```diff
@@ -85,6 +95,18 @@ public void constructor_with_heartbeatTimeout_should_initialize_instance()
...
+        [Fact]
+        public void constructor_with_serverMonitoringMode_should_initialize_instance()
+        {
+            var serverMonitoringMode = ServerMonitoringMode.Stream;
+
+            var subject = new ServerSettings(serverMonitoringMode: serverMonitoringMode);
+
+            subject.HeartbeatInterval.Should().Be(ServerSettings.DefaultHeartbeatInterval);
+            subject.HeartbeatTimeout.Should().Be(ServerSettings.DefaultHeartbeatTimeout);
+            subject.ServerMonitoringMode.Should().Be(ServerMonitoringMode.Stream);
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Core.Tests/Core/Configuration/ServerSettingsTests.cs:107]: We can verify just the `ServerMonitoringMode` here and in `With_serverMonitoringMode_should_return_expected_result`?

File: tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs (hunk 2 above)
Comment (BorisDog) [reply tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs:166]: The spec is not clear about this, so other drivers might have different behavior. I think we should change this in different ticket.

File: tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs [hunk 9]
```diff
@@ -389,6 +415,34 @@ public void ServerMonitor_without_serverApi_but_with_loadBalancedConnection_shou
             sentMessages[0].Should().Be($"{{ opcode : \"opmsg\", requestId : {requestId}, responseTo : 0, exhaustAllowed : true, sections : [ {{ payloadType : 0, document : {{ hello : 1, helloOk: true, topologyVersion : {{ processId : ObjectId(\"000000000000000000000000\"), counter : NumberLong(0) }}, maxAwaitTimeMS : NumberLong(86400000), loadBalanced: true, $db : \"admin\" }} }} ] }}");
         }
 
+        [Fact]
+        public void Should_use_polling_protocol_if_running_in_FaaS_platform()
+        {
+            // set environment to simulate being on FaaS platform
+            Environment.SetEnvironmentVariable("VERCEL", "vercel");
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs:422]: Do we want to test all environment variables?

File: tests/MongoDB.Driver.Tests/UnifiedTestOperations/Matchers/UnifiedEventMatcher.cs (hunk 3 above)
Comment (BorisDog) [reply tests/MongoDB.Driver.Tests/UnifiedTestOperations/Matchers/UnifiedEventMatcher.cs:429]: You could use dynamic, but that's hacky.
Another option is to have `BsonDocument GetHeartbeatDocument(IEvent event)` helper.

File: src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs [hunk 10]
```diff
@@ -471,6 +498,16 @@ private void SetDescription(ServerDescription oldDescription, ServerDescription
             OnDescriptionChanged(oldDescription, newDescription);
         }
 
+        private bool IsRunningInFaaS()
+        {
+            return (Environment.GetEnvironmentVariable("AWS_EXECUTION_ENV")?.StartsWith("AWS_Lambda_") ?? false) ||
+                   Environment.GetEnvironmentVariable("AWS_LAMBDA_RUNTIME_API") != null ||
```
Comment (BorisDog) [inline src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs:504]: Maybe it's worth specifying the FAAS provider in a comment for each var.

File: tests/MongoDB.Driver.Tests/UnifiedTestOperations/UnifiedEntityMap.cs [hunk 11]
```diff
@@ -703,6 +709,8 @@ private IGridFSBucket CreateBucket(BsonDocument entity, Dictionary<string, IMong
                         settings.WriteConcern = writeConcern;
                         settings.HeartbeatInterval = heartbeatFrequency.GetValueOrDefault(defaultValue: TimeSpan.FromMilliseconds(5)); // 5 ms default value for spec tests
                         settings.ServerApi = serverApi;
+                        settings.ServerMonitoringMode =
+                            serverMonitoringMode.GetValueOrDefault(settings.ServerMonitoringMode);
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Tests/UnifiedTestOperations/UnifiedEntityMap.cs:713]: minor: could place assignment in the same line.

**Review by BorisDog — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Core/Core/Servers/RoundTripTimeMonitor.cs [hunk 12]
```diff
@@ -27,6 +27,7 @@ namespace MongoDB.Driver.Core.Servers
     internal interface IRoundTripTimeMonitor : IDisposable
     {
         TimeSpan Average { get; }
+        bool Started { get; }
```
Comment (sanych-sun) [inline src/MongoDB.Driver.Core/Core/Servers/RoundTripTimeMonitor.cs]: _Minor_: should it be named `IsStarted`?

File: src/MongoDB.Driver.Core/Core/Servers/RoundTripTimeMonitor.cs [hunk 13]
```diff
@@ -78,6 +82,17 @@ public TimeSpan Average
             }
         }
 
+        public bool Started
+        {
+            get
+            {
+                lock (_lock)
```
Comment (sanych-sun) [inline src/MongoDB.Driver.Core/Core/Servers/RoundTripTimeMonitor.cs:86]: Are we expecting high concurrency here? If so - it's better to use ReaderWriterLockSlim so to have let multiple readers read the value in parallel.

File: src/MongoDB.Driver.Core/Core/Servers/RoundTripTimeMonitor.cs [hunk 14]
```diff
@@ -99,6 +114,7 @@ public void Start()
         {
             _roundTripTimeMonitorThread = new Thread(ThreadStart) { IsBackground = true };
             _roundTripTimeMonitorThread.Start();
+            _started = true;
```
Comment (sanych-sun) [inline src/MongoDB.Driver.Core/Core/Servers/RoundTripTimeMonitor.cs]: Having lock on read and no-lock on write looks suspicious. We either need both or none.

**Review by sanych-sun — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Legacy/MongoServerSettings.cs [hunk 15]
```diff
@@ -610,6 +613,20 @@ public IEnumerable<MongoServerAddress> Servers
             }
         }
 
+        /// <summary>
+        /// Gets or sets the server monitoring mode.
+        /// </summary>
+        /// <exception cref="System.InvalidOperationException">MongoServerSettings is frozen.</exception>
+        public ServerMonitoringMode ServerMonitoringMode
```
[the remark this answers is not in the document]
Comment (JamesKovacs) [reply src/MongoDB.Driver.Legacy/MongoServerSettings.cs:620]: Excellent point. Customers shouldn’t be using the legacy driver anymore, especially for new applications. FaaS apps are by definition new apps as cloud providers only introduced these capabilities relatively recently. We can remove the changes in legacy.

File: src/MongoDB.Driver.Core/Core/Servers/RoundTripTimeMonitor.cs (hunk 13 above)
Comment (adelinowona) [reply src/MongoDB.Driver.Core/Core/Servers/RoundTripTimeMonitor.cs:86]: I don't think we are expecting high concurrency here, so I will lean towards leaving it as is for now.

Comment (BorisDog) [reply src/MongoDB.Driver.Core/Core/Servers/RoundTripTimeMonitor.cs:86]: Actually we don't need any kind of synchronization at all here.
Proposed refactoring would eliminate the need of this property. Otherwise `Started => _roundTripTimeMonitorThread != null`  would be sufficient.

**Review by BorisDog — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs [hunk 16]
```diff
@@ -178,11 +185,16 @@ public void RequestHeartbeat()
         }
 
         // private methods
+        private bool UsingStreamingProtocol(HelloResult helloResult)
```
Comment (BorisDog) [inline src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs:188]: 1. Alphabetical order.
2. Personal preference : consider short arrow form `=>`
3. Also consider `IsUsingStreamingProtocol` as just `Using` might be read as some action.

File: src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs [hunk 17]
```diff
@@ -405,13 +417,19 @@ private void Heartbeat(CancellationToken cancellationToken)
...
                 }
 
-                processAnother =
-                    // serverSupportsStreaming
-                    (newDescription.Type != ServerType.Unknown && heartbeatHelloResult != null && heartbeatHelloResult.TopologyVersion != null) ||
-                    // connectionIsStreaming
-                    (helloProtocol != null && helloProtocol.MoreToCome) ||
-                    // transitionedWithNetworkError
-                    (IsNetworkError(heartbeatException) && previousDescription.Type != ServerType.Unknown);
+                var serverSupportsStreaming = newDescription.Type != ServerType.Unknown &&
```
Comment (BorisDog) [inline src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs:420]: Alignment of new lines should be: 
```
var serverSupportsStreaming = newDescription.Type != ServerType.Unknown &&
   heartbeatHelloResult != null &&
   heartbeatHelloResult.TopologyVersion != null;
```

File: src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs [hunk 18]
```diff
@@ -471,6 +495,22 @@ private void SetDescription(ServerDescription oldDescription, ServerDescription
             OnDescriptionChanged(oldDescription, newDescription);
         }
 
+        private bool IsRunningInFaaS()
```
Comment (BorisDog) [inline src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs:498]: Place with `IsUsing`, in alphabetical order.

File: src/MongoDB.Driver/MongoDB.Driver.csproj [hunk 19]
```diff
@@ -14,8 +14,8 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="MongoDB.Libmongocrypt" Version="1.8.0" />
     <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="2.0.0" />
+    <PackageReference Include="MongoDB.Libmongocrypt" Version="1.8.0" />
```
Comment (BorisDog) [inline src/MongoDB.Driver/MongoDB.Driver.csproj:18]: unrelated change.

File: tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs [hunk 20]
```diff
@@ -293,14 +293,16 @@ public void InitializeHelloProtocol_should_use_streaming_protocol_when_available
...
             var subject = CreateSubject(out var mockConnection, out _, out var mockRoundTripTimeMonitor);
 
-            SetupHeartbeatConnection(mockConnection);
+            SetupHeartbeatConnection(mockConnection, isStreamable: true, false);
+            mockConnection.EnqueueCommandResponseMessage(CreateHeartbeatCommandResponseMessage(true), null);
+            mockConnection.EnqueueCommandResponseMessage(CreateHeartbeatCommandResponseMessage(), null);
 
             subject.Initialize();
-            subject.Initialize();
+            SpinWait.SpinUntil(() => mockConnection.GetSentMessages().Count >= 1, TimeSpan.FromSeconds(5)).Should().BeTrue();
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs:305]: I don't understand how this test works, let's sync offline.

File: tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs [hunk 21]
```diff
@@ -327,6 +329,29 @@ public void RequestHeartbeat_should_force_another_heartbeat()
...
+            mockConnection.EnqueueCommandResponseMessage(CreateHeartbeatCommandResponseMessage());
+
+            subject.Initialize();
+            SpinWait.SpinUntil(() => mockConnection.GetSentMessages().Count >= 2, TimeSpan.FromSeconds(5)).Should().BeTrue();
+
+            capturedEvents.Next().Should().BeOfType<ServerHeartbeatStartedEvent>().Subject.Awaited.Should().Be(false);
+            capturedEvents.Next().Should().BeOfType<ServerHeartbeatSucceededEvent>().Subject.Awaited.Should().Be(false);
+            capturedEvents.Next().Should().BeOfType<ServerHeartbeatStartedEvent>().Subject.Awaited.Should().Be(false);
+            capturedEvents.Next().Should().BeOfType<ServerHeartbeatSucceededEvent>().Subject.Awaited.Should().Be(false);
+        }
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs:353]: It this stable?
Waiting for >= 2 events, but checking for 4.

File: tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs [hunk 22]
```diff
@@ -389,6 +414,40 @@ public void ServerMonitor_without_serverApi_but_with_loadBalancedConnection_shou
...
+
+            subject.Initialize();
+            SpinWait.SpinUntil(() => mockConnection.GetSentMessages().Count >= 2, TimeSpan.FromSeconds(5)).Should().BeTrue();
+
+            capturedEvents.Next().Should().BeOfType<ServerHeartbeatStartedEvent>().Subject.Awaited.Should().Be(false);
+            capturedEvents.Next().Should().BeOfType<ServerHeartbeatSucceededEvent>().Subject.Awaited.Should().Be(false);
+            capturedEvents.Next().Should().BeOfType<ServerHeartbeatStartedEvent>().Subject.Awaited.Should().Be(false);
+            capturedEvents.Next().Should().BeOfType<ServerHeartbeatSucceededEvent>().Subject.Awaited.Should().Be(false);
+
+            Environment.SetEnvironmentVariable(environmentVariableName, null);
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs:448]: try/finally. Or even better move `DisposableEnvironmentVariable` to utils and reuse it.

**Review by BorisDog — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs [hunk 23]
```diff
@@ -427,22 +445,28 @@ bool IsNetworkError(Exception ex)
...
 
             var stopwatch = Stopwatch.StartNew();
             try
             {
                 var helloResult = HelloHelper.GetResult(connection, helloProtocol, cancellationToken);
                 stopwatch.Stop();
 
-                _eventLoggerSdam.LogAndPublish(new ServerHeartbeatSucceededEvent(connection.ConnectionId, stopwatch.Elapsed, connection.Description.HelloResult.TopologyVersion != null, helloResult.Wrapped));
+                // RTT check if using polling monitoring
+                if (UsingStreamingProtocol(connection.Description.HelloResult) is not true)
```
Comment (sanych-sun) [inline src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs:457]: Minor: why do we need pattern matching here? Probably could be simplified with:
`if (!UsingStreamingProtocol(connection.Description.HelloResult))`

**Review by sanych-sun — APPROVED** [review]
LGTM + minor comment

File: tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs (hunk 21 above)
Comment (adelinowona) [reply tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs:353]: Waiting for the 2 sent messages should generate the 4 heartbeat events. Indeed it was a bit unstable and I had already fixed it locally. I just have to push the changes. Basically now I am straight up waiting for the 4 events to be ready before proceeding.

File: tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs (hunk 20 above)
Comment (adelinowona) [reply tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs:305]: Discussed offline.

File: tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs [hunk 24]
```diff
@@ -470,7 +565,7 @@ private CommandResponseMessage CreateHeartbeatCommandResponseMessage(bool moreTo
                     new BsonDocument
                     {
                         { "processId", ObjectId.Parse("5ee3f0963109d4fe5e71dd28") },
-                        { "counter", 0 }
+                        { "counter", new BsonInt64(0) }
```
Comment (adelinowona) [inline tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs:568]: This was a bug as the `Topology` get method from `HelloResult` expects this to be a 64-bit int and thus silently fails omitting the topology version even though it was provided in the response. This caused my tests to not behave properly even when I knew the server supported streaming based on the responses.

Comment (BorisDog) [reply tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs:569]: nice find

File: tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs [hunk 25]
```diff
@@ -300,18 +300,51 @@ public void InitializeHelloProtocol_should_use_streaming_protocol_when_available
...
 
             mockRoundTripTimeMonitor.Verify(m => m.Start(), Times.Once);
         }
 
+        [Fact]
+        public void RoundTripTimeMonitor_should_not_be_started_if_using_polling_protocol()
+        {
+            var serverMonitorSettings = new ServerMonitorSettings(
+                TimeSpan.FromSeconds(5),
+                TimeSpan.FromMilliseconds(100),
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs:331]: Can we try even lower heartbeat interval, like 5ms in all tests?

File: tests/MongoDB.Driver.Core.TestHelpers/DisposableEnvironmentVariable.cs [hunk 26]
```diff
@@ -0,0 +1,20 @@
+using System;
+
+namespace MongoDB.Driver.Core.TestHelpers;
+
+public class DisposableEnvironmentVariable : IDisposable
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Core.TestHelpers/DisposableEnvironmentVariable.cs:5]: nice.
make is sealed for Bon tone.

File: tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs [hunk 27]
```diff
@@ -1,17 +1,17 @@
...
-*
-* http://www.apache.org/licenses/LICENSE-2.0
-*
-* Unless required by applicable law or agreed to in writing,.Setup software
-* distributed under the License is distributed on an "AS IS" BASIS,
-* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-* See the License for the specific language governing permissions and
-* limitations under the License.
-*/
+ *
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs:2]: revert?

**Review by BorisDog — CHANGES_REQUESTED** [review]

File: tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs [hunk 28]
```diff
@@ -300,18 +300,51 @@ public void InitializeHelloProtocol_should_use_streaming_protocol_when_available
...
 
             subject.Initialize();
-            subject.Initialize();
+            // spin until the initial handshake has happened in which case the server description state should be Connected
+            SpinWait.SpinUntil(() => subject.Description.State == ServerState.Connected, TimeSpan.FromSeconds(5)).Should().BeTrue();
+
+            // RTT thread should be started after first heartbeat (initial handshake)
+            mockRoundTripTimeMonitor.Verify(m => m.Start(), Times.Once);
+
+            SpinWait.SpinUntil(() => capturedEvents.Count >= 4, TimeSpan.FromSeconds(5)).Should().BeTrue();
```
[the remark this answers is not in the document]
Comment (adelinowona) [reply tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerMonitorTests.cs:321]: `IsStarted` is setup in the `CreateSubject` method, it switches to returning true after it is called once. I will add checking `IsStarted` call count as well.

**Review by BorisDog — APPROVED** [review]
LGTM! + minor comment

File: tests/MongoDB.Driver.Tests/UnifiedTestOperations/UnifiedEntityMap.cs [hunk 29]
```diff
@@ -18,13 +18,15 @@
 using System.Collections.Generic;
 using System.Linq;
 using System.Threading.Tasks;
+using FluentAssertions;
```
[the remark this answers is not in the document]
Comment (adelinowona) [reply tests/MongoDB.Driver.Tests/UnifiedTestOperations/UnifiedEntityMap.cs]: Nope I'll remove it!
